src/merkle_tree.ml

```diff
@@ -426,19 +437,24 @@ struct
     go 0 entry_hash addr0 path0
 
   type _ Request.t +=
-    | Get_element : Address.value -> (Elt.value * Path.value) Request.t
+    | Get_element :
+        Tag.value * Address.value
+        -> (Elt.value * Path.value) Request.t
     | Get_path : Address.value -> Path.value Request.t
     | Set : Address.value * Elt.value -> unit Request.t
 
   (* addr0 should have least significant bit first *)
-  let%snarkydef_ modify_req ~(depth : int) root addr0 ~f :
-      (Hash.var, 's) Checked.t =
+  let%snarkydef_ modify_req ~is_chain_voting ~(depth : int) root addr0 ~f :
+      (Hash.var * Elt.var, 's) Checked.t =
     let open Let_syntax in
     let%bind prev, prev_path =
       request_witness
         Typ.(Elt.typ * Path.typ ~depth)
         As_prover.(
-          map (read (Address.typ ~depth) addr0) ~f:(fun a -> Get_element a))
+          Let_syntax.(
+            let%map addr = read (Address.typ ~depth) addr0
+            and tag = read Tag.typ is_chain_voting in
```

Comment:
This `read` would be better living outside the code for this and being more closely integrated with the handler. I would feel much more comfortable with something like:

```ocaml
let%bind tag =
  As_prover.(Ref.create (read Tag.typ is_chain_voting))
in
let code = (* code that calls modify_ref *) in
let%bind (* variables here *) =
  handle code As_prover.(fun (With {request; respond}) ->
    match request with
    | Get_element addr ->
        let%bind tag = Ref.get tag in
        (* handler code... *)
    | _ -> unhandled)
in
(* more code *)
```
at the callsite and removing this completely.

Reply:
I agree with you. I will put the logic inside coda then.

Reply:
I am confused about this code snippet. I thought the second argument to `handle` should be a `Handler.t` . But `let%bind tag = Ref.get tag in` would be of type `As_prover.t`.

I am still learning the basics of snarky. Would you please explain what's happening here?

Reply:
Sorry, I got my wires a bit crossed on this. I think the code has to be

```ocaml
let tag = ref None in
let%bind tag =
  as_prover As_prover.(
    let%map is_chain_voting = read Tag.typ is_chain_voting in
    tag := Some chain_voting
  )
in
let code = (* code that calls modify_ref *) in
let%bind (* variables here *) =
  handle code As_prover.(fun (With {request; respond}) ->
    match request with
    | Get_element addr ->
        let tag = Option.value_exn !tag in
        (* handler code... *)
    | _ -> unhandled)
in
(* more code *)
```

I think there's probably value in us being able to pass `As_prover` blocks as handlers, so I'll make a PR for that tomorrow.

Reply:
This works. Thanks a lot.